vault/resource_ssh_secret_backend_role.go

```diff
@@ -246,32 +340,106 @@ func sshSecretBackendRoleRead(d *schema.ResourceData, meta interface{}) error {
 		d.SetId("")
 		return nil
 	}
-	d.Set("name", name)
-	d.Set("backend", backend)
-	d.Set("key_type", role.Data["key_type"])
-	d.Set("allow_bare_domains", role.Data["allow_bare_domains"])
-	d.Set("allow_host_certificates", role.Data["allow_host_certificates"])
-	d.Set("allow_subdomains", role.Data["allow_subdomains"])
-	d.Set("allow_user_certificates", role.Data["allow_user_certificates"])
-	d.Set("allow_user_key_ids", role.Data["allow_user_key_ids"])
-	d.Set("allowed_critical_options", role.Data["allowed_critical_options"])
-	d.Set("allowed_domains", role.Data["allowed_domains"])
-	d.Set("cidr_list", role.Data["cidr_list"])
-	d.Set("allowed_extensions", role.Data["allowed_extensions"])
-	d.Set("default_extensions", role.Data["default_extensions"])
-	d.Set("default_critical_options", role.Data["default_critical_options"])
-	d.Set("allowed_users_template", role.Data["allowed_users_template"])
-	d.Set("allowed_users", role.Data["allowed_users"])
-	d.Set("default_user", role.Data["default_user"])
-	d.Set("key_id_format", role.Data["key_id_format"])
-	d.Set("allowed_user_key_lengths", role.Data["allowed_user_key_lengths"])
-	d.Set("max_ttl", role.Data["max_ttl"])
-	d.Set("ttl", role.Data["ttl"])
-	d.Set("algorithm_signer", role.Data["algorithm_signer"])
+
+	if err := d.Set("name", name); err != nil {
+		return err
+	}
+
+	if err := d.Set("backend", backend); err != nil {
+		return err
+	}
+
+	fields := []string{
+		"key_type", "allow_bare_domains", "allow_host_certificates",
+		"allow_subdomains", "allow_user_certificates", "allow_user_key_ids",
+		"allowed_critical_options", "allowed_domains",
+		"cidr_list", "allowed_extensions", "default_extensions",
+		"default_critical_options", "allowed_users_template",
+		"allowed_users", "default_user", "key_id_format",
+		"max_ttl", "ttl", "algorithm_signer",
+	}
+
+	for _, k := range fields {
+		if err := d.Set(k, role.Data[k]); err != nil {
+			return err
+		}
+	}
+
+	if err := setSSHRoleKeyConfig(d, role); err != nil {
+		return err
+	}
 
 	return nil
 }
 
+func setSSHRoleKeyConfig(d *schema.ResourceData, role *api.Secret) error {
+	keyConfigs, err := getSSHRoleKeyConfig(role)
+	if err != nil {
+		return err
+	}
+
+	newField := "allowed_user_key_lengths"
+	legacyField := "allowed_user_key_lengths"
+	// work around to support the allowed_user_key_lengths with Vault 1.10+
+	if _, ok := d.GetOk(legacyField); ok {
+		v := make(map[string]interface{})
+		for _, val := range keyConfigs {
+			keyType := val["type"].(string)
+			lengths := val["lengths"].([]interface{})
+			var l interface{}
+			count := len(lengths)
+			if count > 0 {
+				l = lengths[0]
+			}
+
+			if count > 1 {
+				log.Printf("[WARN] Vault 1.10+ returned more than one "+
+					"key length for key type %q, specify key lengths in a "+
+					"%q block instead of %q with values %v",
+					newField, keyType, legacyField, lengths)
+			}
+
+			v[keyType] = l
+		}
+
+		return d.Set(newField, v)
```

Comment:
Since this is in the case the user set the `legacyField` in the TF config, should we be setting `legacyField` to `v` in the TF state instead?

Reply:
good catch, will update per your other comment.

Reply:
Fixed in https://github.com/hashicorp/terraform-provider-vault/pull/1413/commits/f724d6aa18390fa25bc09807b2491507349d6ea1

Reply:
Fixed in https://github.com/hashicorp/terraform-provider-vault/pull/1413/commits/affd6a143fb54c76bd002c2dbd24f632ffe477e6